File: Problem1_2Dmatrix.py
'''
Time Complexity: O(logmn)
Space Complexity: O(1)
Implemented Successfully
'''

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if(matrix == None or len(matrix) == 0):
            return False
        m, n = len(matrix), len(matrix[0])
        low, high = 0, (m*n)-1
        while (low <= high):
            mid = low + (high - low)//2
            row = mid//n
            col = mid%n
            if(matrix[row][col] == target):
                return True
            elif(matrix[row][col] < target):
                low = mid+1
            elif(matrix[row][col] > target):
                high = mid-1
            else:
                return False

File: Problem1-2Dmatrix.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length == 0) return false;
        int m = matrix.length;    //no of columns
        int n = matrix[0].length; //no of rows

        int low = 0;
        int high = m*n - 1; // since we are flattening it out

        while(low <= high){
            int mid = low + (high-low)/2;
            int r = mid/n;
            int c = mid%n;
            
            if (matrix[r][c] == target) return true;

            else if (matrix[r][c] > target){
                high = mid -1;        
            }

            else {
                low = mid +1;
            }

        }
        return false;
    }
}